### PR TITLE
Adding functional tests

### PR DESCRIPTION
**Pull request by cloudnull**
Signed-off-by: Kevin Carter <kevin.carter@rackspace.com>

### CODE REVIEW

**Comment by evrardjp** [comment]
Too many things done at the same time, easy to be lost...

File: defaults/main.yml [hunk 1]
```diff
@@ -585,8 +585,6 @@ openrc_insecure: "{{ (keystone_service_adminuri_insecure | bool or keystone_serv
 # Name of the virtual env to deploy into (maas)
 maas_venv: "/openstack/venvs/maas-{{ rpc_release }}"
 maas_venv_bin: "{{ maas_venv }}/bin"
-# Set this to enable or disable installing in a venv
-maas_venv_enabled: true
```
Comment (evrardjp) [inline defaults/main.yml]: Why is this removed? Why is this removed in the same commit?

File: run-tests.sh [hunk 2]
```diff
@@ -42,34 +42,9 @@ if which yum; then
...
-      --assume-yes install $BINDEP_PKGS
-elif which yum; then
-    # Don't run yum with an empty list of packages.
-    # It will fail and cause the script to exit with an error.
-    if [[ ${#BINDEP_PKGS} > 0 ]]; then
-      sudo yum install -y $BINDEP_PKGS
-    fi
+if [ "${FUNCTIONAL_TEST}" = true ]; then
+  tox -e bindep
+  tox -e functional
```
Comment (evrardjp) [inline run-tests.sh]: This is clear, I love it. I'd rather do that into a separate commit though, to be able to reapply everywhere else in OSA.

File: run_tests.sh [hunk 3]
```diff
@@ -1,75 +0,0 @@
-#!/usr/bin/env bash
```
Comment (evrardjp) [inline run_tests.sh]: I like the cleanup of the run_tests.sh/run-tests.sh. We should have only one way to do it (TBD), and apply everywhere. Want help there?

File: tests/group_vars/all.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+physical_host: localhost
+
+# This is used for RPC specific things and is required
+# NOTE(cloudnull): Before this was hardcoded, now it will look at the playbook path and
+#                  take the git information from the root playbook directory dynamically.
+rpc_release: "{{ lookup('pipe', 'cd ' + playbook_dir.split('common')[0] + '; git describe --tags --abbrev=0; cd -').splitlines()[0].strip() }}"
```
Comment (evrardjp) [inline tests/group_vars/all.yml]: This may have to change into a static thing. Could we instead checkout rpc-openstack and get the value from there? (We have a script for that already).

File: tests/ansible-role-requirements.yml [hunk 5]
```diff
@@ -1,5 +1,76 @@
...
+  version: stable/newton
+- name: lxc_hosts
+  src: https://git.openstack.org/openstack/openstack-ansible-lxc_hosts
+  scm: git
+  version: stable/newton
 - name: pip_install
   src: https://git.openstack.org/openstack/openstack-ansible-pip_install
   scm: git
-  version: master
+  version: stable/newton
```
Comment (evrardjp) [inline tests/ansible-role-requirements.yml]: This line makes sense.

File: tests/ansible-role-requirements.yml [hunk 6]
```diff
@@ -1,5 +1,76 @@
...
+- name: rabbitmq_server
```
Comment (evrardjp) [inline tests/ansible-role-requirements.yml]: All the rest, should, IMO go away. 
What are the functional elements we should do? Here you're trying to integrate rpc-maas into OSA, which is fine for me, but it seems carrying a lot of burden: We already have an RPC-O job that does exactly this. We should just change the a-r-r there to point at master of rpc-maas.

**Comment by evrardjp** [comment]
Same error as I have here:

which is:

```
TASK [rpc-maas : Assign agent ID to entity] ************************************
task path: /var/lib/jenkins/workspace/RPC-IRR_rpc_maas-master-trusty-base/rcbops/rpc-maas/tasks/agent_install_remote.yml:16
skipping: [infra1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
<10.1.1.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<10.1.1.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721 `" && echo ansible-tmp-1493507025.49-248049955277721="` echo $HOME/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721 `" ) && sleep 0'
<10.1.1.1> PUT /tmp/tmp2ZW20Z TO /root/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721/raxmon
<10.1.1.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721/ /root/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721/raxmon && sleep 0'
<10.1.1.1> EXEC /bin/sh -c 'LC_MESSAGES=en_US.UTF-8 LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 /usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721/raxmon; rm -rf "/root/.ansible/tmp/ansible-tmp-1493507025.49-248049955277721/" > /dev/null 2>&1 && sleep 0'
skipping: [infra2] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [infra3] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [openstack1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [swift1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_1QzptR/ansible_module_raxmon.py", line 173, in <module>
    main()
  File "/tmp/ansible_1QzptR/ansible_module_raxmon.py", line 164, in main
    conn = _get_conn(get_driver, Provider, module.params['raxmon_cfg'])
  File "/tmp/ansible_1QzptR/ansible_module_raxmon.py", line 84, in _get_conn
    conn = driver(user, api_key)
  File "/var/lib/jenkins/workspace/RPC-IRR_rpc_maas-master-trusty-base/rcbops/rpc-maas/.tox/functional/lib/python2.7/site-packages/rackspace_monitoring/drivers/rackspace.py", line 173, in __init__
    self._initialize_connection_base_url(constructor_kwargs=kwargs)
  File "/var/lib/jenkins/workspace/RPC-IRR_rpc_maas-master-trusty-base/rcbops/rpc-maas/.tox/functional/lib/python2.7/site-packages/rackspace_monitoring/drivers/rackspace.py", line 187, in _initialize_connection_base_url
    self.connection._populate_hosts_and_request_paths()
  File "/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack.py", line 327, in _populate_hosts_and_request_paths
    osa = osa.authenticate(**kwargs)  # may throw InvalidCreds
  File "/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack_identity.py", line 853, in authenticate
    return self._authenticate_2_0_with_api_key()
  File "/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack_identity.py", line 868, in _authenticate_2_0_with_api_key
    return self._authenticate_2_0_with_body(reqbody)
  File "/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack_identity.py", line 888, in _authenticate_2_0_with_body
    raise InvalidCredsError()
libcloud.common.types.InvalidCredsError: 'Invalid credentials with the provider'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "raxmon"}, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_1QzptR/ansible_module_raxmon.py\", line 173, in <module>\n    main()\n  File \"/tmp/ansible_1QzptR/ansible_module_raxmon.py\", line 164, in main\n    conn = _get_conn(get_driver, Provider, module.params['raxmon_cfg'])\n  File \"/tmp/ansible_1QzptR/ansible_module_raxmon.py\", line 84, in _get_conn\n    conn = driver(user, api_key)\n  File \"/var/lib/jenkins/workspace/RPC-IRR_rpc_maas-master-trusty-base/rcbops/rpc-maas/.tox/functional/lib/python2.7/site-packages/rackspace_monitoring/drivers/rackspace.py\", line 173, in __init__\n    self._initialize_connection_base_url(constructor_kwargs=kwargs)\n  File \"/var/lib/jenkins/workspace/RPC-IRR_rpc_maas-master-trusty-base/rcbops/rpc-maas/.tox/functional/lib/python2.7/site-packages/rackspace_monitoring/drivers/rackspace.py\", line 187, in _initialize_connection_base_url\n    self.connection._populate_hosts_and_request_paths()\n  File \"/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack.py\", line 327, in _populate_hosts_and_request_paths\n    osa = osa.authenticate(**kwargs)  # may throw InvalidCreds\n  File \"/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack_identity.py\", line 853, in authenticate\n    return self._authenticate_2_0_with_api_key()\n  File \"/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack_identity.py\", line 868, in _authenticate_2_0_with_api_key\n    return self._authenticate_2_0_with_body(reqbody)\n  File \"/openstack/venvs/maas-11.0.0rc1/lib/python2.7/site-packages/libcloud/common/openstack_identity.py\", line 888, in _authenticate_2_0_with_body\n    raise InvalidCredsError()\nlibcloud.common.types.InvalidCredsError: 'Invalid credentials with the provider'\n", "module_stdout": "", "msg": "MODULE FAILURE"}
```

i.e.: our gates should have the proper credentials I guess.

File: defaults/main.yml (hunk 1 above)
Comment (prometheanfire) [inline defaults/main.yml]: I'm thinking it's because of holland being put in a separate venv, but dunno, an answer here would be nice.

File: tasks/ceph.yml [hunk 7]
```diff
@@ -17,18 +17,24 @@
   ceph_osd_host_facts:
   register: ceph_osd_facts
   tags: always
-  when: inventory_hostname in groups['osds']
+  when:
+    - "'osds' in groups"
+    - groups['osds'] | length > 0
+    - inventory_hostname in groups['osds'] | default([])
```
Comment (prometheanfire) [inline tasks/ceph.yml]: lol, safety

File: tests/ansible-role-requirements.yml [hunk 8]
```diff
@@ -1,5 +1,76 @@
----
```
Comment (prometheanfire) [inline tests/ansible-role-requirements.yml]: first line shouldn't be removed, just corrected to `---` from `----`

Comment (cloudnull) [inline tests/ansible-role-requirements.yml]: done.

File: tests/ansible-role-requirements.yml (hunk 6 above)
Comment (cloudnull) [inline tests/ansible-role-requirements.yml]: I'm building the full test which is akin to what we had before, in future PRs we can pick it apart and do more focused testing.

File: defaults/main.yml (hunk 1 above)
Comment (cloudnull) [inline defaults/main.yml]: ++ mass is in a venv now, which is what was happening in the gate anyway.

File: tasks/ceph.yml (hunk 7 above)
Comment (cloudnull) [inline tasks/ceph.yml]: with this we can start gating pieces individually.not that we are atm but soon.

File: defaults/main.yml [hunk 9]
```diff
@@ -613,7 +611,7 @@ verify_maas_retries: 5
 # maas_use_api: Allow operations that make use of the MaaS api, set to false
 #               for offline testing
 #
-maas_use_api: true
+maas_use_api: false
```
Comment (alextricity25) [inline defaults/main.yml]: Setting this to false here means that we'll have to override it to true in ``user_rpco_variables_defaults.yml``. Otherwise RPCO deployments won't register checks with the MaaS API by default.

File: defaults/main.yml (hunk 1 above)
Comment (alextricity25) [inline defaults/main.yml]: :+1:

File: defaults/main.yml [hunk 10]
```diff
@@ -634,3 +632,13 @@ swift_accesscheck_file: "index.html"
...
 swift_service_project_name: "service"
+
+# Instruct the swift service checks to be ldap aware.
+swift_service_in_ldap: false
+
+# Holland venv enabled
+holland_venv_enabled: true
+holland_venv_bin: "/openstack/venvs/holland-{{ rpc_release }}/bin"
+
+# Set the openstack-release to testing by default
```
Comment (alextricity25) [inline defaults/main.yml]: Are you sure this needs to be in ``defaults/main.yml``? It almost seems to be like this variable needs to be in the ``rpc-maas-overrides.yml``  file under the ``tests`` directory

File: tests/inventory [hunk 11]
```diff
@@ -1,26 +1,199 @@
+#### ALL HOSTS AND CONTAINERS START HERE
+
 [all]
-localhost physical_host=localhost
+localhost ansible_become=True ansible_connection=local ansible_user=root
+infra1
```
Comment (alextricity25) [inline tests/inventory]: Are infra2 and infra3 also suppose to be here?

**Comment by alextricity25** [comment]
It might be a good idea to bring in verify-maas.yml and run that playbook as part of the testing suite you are building here.
https://github.com/rcbops/rpc-openstack/blob/master/rpcd/playbooks/verify-maas.yml

**Comment by cloudnull** [comment]
@alextricity25 absolutely. This PR is growing and now that we have a functional convergence I'd like to cap off the changes here, get this merged, and then continue making additional tests.

File: tests/inventory (hunk 11 above)
Comment (cloudnull) [inline tests/inventory]: no just 1 infra, the rpc-o integrated gate can test this with >1 node. That said, the next act will be to break down the actual testing being done and create a matrix of tests that can be run individually.

File: defaults/main.yml (hunk 10 above)
Comment (cloudnull) [inline defaults/main.yml]: There needs to be a default set for the option otherwise independent execution will fail because the option is undefined.

File: defaults/main.yml (hunk 9 above)
Comment (cloudnull) [inline defaults/main.yml]: correct.

File: tests/inventory (hunk 11 above)
Comment (alextricity25) [inline tests/inventory]: Gotcha. The reason I asked is because you have ``tests/host_vars/infra2.yml`` and ``tests/host_vars/infra3.yml`` in this PR

File: defaults/main.yml (hunk 9 above)
Comment (evrardjp) [inline defaults/main.yml]: Then we need a release note and a relative change in rpc-o.

File: defaults/main.yml (hunk 10 above)
Comment (evrardjp) [inline defaults/main.yml]: I kinda agree on the failure by default, and have the tests/rpc-maas-overrides.yml to testing

We could even have a

```
assert:
  that:
    - "openstack_release is defined"
```

File: tests/host_vars/openstack1.yml [hunk 12]
```diff
@@ -28,4 +28,4 @@ neutron_local_ip: "{{ ansible_host }}"
 
 swift_replication_address: 10.4.1.101
 
-swift_storage_address: "{{ storage_address }}"
+swift_storage_address: "{{ storage_address }}"
```
Comment (evrardjp) [inline tests/host_vars/openstack1.yml]: at some point you should change your text editor to not have these :p 

POSIX!

File: tests/rpc-maas-overrides.yml [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+nova_venv_tag: "testing"
+sahara_venv_tag: "testing"
+swift_venv_tag: "testing"
+
+maas_rabbitmq_password: secrete
+
+# This is used for RPC specific things and is required
+# NOTE(cloudnull): Before this was hardcoded, now it will look at the playbook path and
+#                  take the git information from the root playbook directory dynamically.
+rpc_release: "{{ lookup('pipe', 'cd ' + playbook_dir.split('common')[0] + '; git describe --tags --abbrev=0; cd -').splitlines()[0].strip() }}"
```
Comment (evrardjp) [inline tests/rpc-maas-overrides.yml]: I'd add openstack_release here too.

File: tests/rpc_maas-overrides.yml [hunk 14]
```diff
@@ -0,0 +1 @@
+rpc-maas-overrides.yml
```
Comment (evrardjp) [inline tests/rpc_maas-overrides.yml]: I'd rather remove the symlink and use the standard 
```<role name>-overrides.yml```

File: tasks/ceph.yml (hunk 7 above)
Comment (evrardjp) [inline tasks/ceph.yml]: The last one is ambiguous without parenthesis. Please add some for readability.

File: tasks/maas_host_monitoring.yml [hunk 15]
```diff
@@ -13,15 +13,19 @@
...
-  roles:
-    - role: "{{ horizon_rolename | default('os_horizon') }}"
-  vars_files:
-    - common/test-vars.yml
\ No newline at end of file
+- include: cdm.yml
+
+# NOTE(Cloudnull): This will load only the one vendor specific host file
+#  and fail if not found when host_check is true
+- include: "{{ ansible_system_vendor.split()[0] | lower }}.yml"
```
Comment (evrardjp) [inline tasks/maas_host_monitoring.yml]: Are you sure the facts are gathered beforehand?

Else we need a setup task.

Also that would be good to explain why you split() [on all whitespaces?] instead of split(" "), for posterity.

File: tests/rpc-maas-overrides.yml [hunk 16]
```diff
@@ -13,54 +13,82 @@
...
-ironic_venv_tag: "testing"
-keystone_venv_tag: "testing"
-neutron_venv_tag: "testing"
-nova_venv_tag: "testing"
-sahara_venv_tag: "testing"
-swift_venv_tag: "testing"
+bridges:
+  - name: "br-mgmt"
+    ip_addr: "10.1.1.1"
+  - name: "br-vxlan"
```
Comment (evrardjp) [inline tests/rpc-maas-overrides.yml]: do we need all these nets?

File: defaults/main.yml (hunk 1 above)
Comment (evrardjp) [inline defaults/main.yml]: Well then we need to remove the var : p
